typing/ctype.ml

```diff
@@ -2525,6 +2788,66 @@ let find_lowest_level ty =
 let find_expansion_scope env path =
   (Env.find_type path env).type_expansion_scope
 
+let layout_of_abstract_type_declaration env p =
+  try
+    (* This lookup duplicates work already done in is_instantiable, which guards
+       the case of unify3 that reaches this function.  Would be nice to
+       eliminate the duplication, but is seems tricky to do so without
+       complicating unify3. *)
+    match (Env.find_type p env).type_kind with
+    | Type_abstract {layout} -> layout
+    | _ -> assert false
+  with
+    Not_found -> assert false
+
+let add_layout_equation env destination layout1 =
+  (* Here we check whether the source and destination layouts intersect.  If
+     they don't, we can give a type error.  If they do, and destination is
+     abstract, we can improve type checking by assigning destination that
+     layout.  If destination is not abstract, we can't locally improve its
+     layout, so we're slightly incomplete.  *)
```

Comment:
In your example, I think we can agree it's fine to define `destination` to be immediate.  It feels a little odd, but we're in dead code here (because you can't prove the equality we'd need to be in a match on), so it's sound, and it's not that unusual to have code under gadt matches that can't actually happen.

The next question is, do we need to?  And I think the answer is yes.  We need to refine the layout of destination _sometimes_ (recall our immediate64 examples) and we don't have a way (now) to distinguish "value" from "non-immediate value" as in your example.  Even if we did, I don't think we could use it here, because the imm64 case is exactly a safe refinement of a type that was previously though to be non-immediate to an immediate64? For the same reason, I don't think Stephen's idea works, and I'm deleting the CR.  And anyway, it feels in keeping with the spirit of gadts to allow matching on false.